internal/terraform/workspace/workspace.go

```diff
@@ -0,0 +1,284 @@
+	if err != nil {
+		return err
+	}
+
+	version, err := tf.Version()
+	if err != nil {
+		return err
+	}
+
+	err = schema.SchemaSupportsTerraform(version)
+	if err != nil {
+		return err
+	}
+
+	p, err := lang.FindCompatibleParser(version)
+	if err != nil {
+		return err
+	}
+	p.SetLogger(w.logger)
+
+	ss := w.newSchemaStorage()
+
+	ss.SetLogger(w.logger)
+
+	p.SetSchemaReader(ss)
+
+	w.parser = p
+	w.schemaWriter = ss
+	w.tfExec = tf
+	w.tfVersion = version
+
+	err = w.initPluginCache(dir)
+	if err != nil {
+		return fmt.Errorf("plugin initialization failed: %w", err)
+	}
+	err = w.initModuleCache(dir)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (w *workspace) initTfExecutor(dir string) (*exec.Executor, error) {
+	tfPath := w.tfExecPath
+	if tfPath == "" {
+		var err error
+		tfPath, err = w.tfDiscoFunc()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	tf := w.tfNewExecutor(w.ctx, tfPath)
+
+	tf.SetWorkdir(dir)
+	tf.SetLogger(w.logger)
+
+	if w.tfExecLogPath != "" {
+		tf.SetExecLogPath(w.tfExecLogPath)
+	}
+
+	if w.tfExecTimeout != 0 {
+		tf.SetTimeout(w.tfExecTimeout)
+	}
+
+	return tf, nil
+}
+
+func (w *workspace) initPluginCache(dir string) error {
+	var lf *TrackedFile
+	if w.ignorePluginCache {
+		lf = &TrackedFile{
+			Path: pluginLockFilePaths(dir)[0],
+		}
+	} else {
+		var err error
+		lockPaths := pluginLockFilePaths(dir)
+		lf, err = findTrackedFile(lockPaths)
+		if err != nil {
+			if os.IsNotExist(err) {
+				return &tferr.NotInitializedErr{Dir: dir}
+			}
+
+			return fmt.Errorf("unable to calculate hash: %w", err)
+		}
+	}
+
+	return w.UpdatePluginCache(lf)
+}
+
+func (w *workspace) initModuleCache(dir string) error {
+	lf, err := trackedFileFromPath(moduleLockFilePath(dir))
+	if err != nil {
+		if os.IsNotExist(err) {
+			w.logger.Printf("module lock file not found: %s", err.Error())
+			return nil
+		}
+
+		return fmt.Errorf("unable to calculate hash: %w", err)
+	}
+
+	return w.UpdateModuleCache(lf)
+}
+
+func (w *workspace) UpdateModuleCache(lockFile *TrackedFile) error {
+	w.moduleMu.Lock()
+	w.logger.Printf("updating module cache based on %s ...", lockFile.Path)
+	defer w.moduleMu.Unlock()
+
+	w.moduleLockFile = lockFile
+
+	b, err := ioutil.ReadFile(lockFile.Path)
+	if err != nil {
+		return err
+	}
+	mc := ModuleCache{}
+	err = json.Unmarshal(b, &mc)
+	if err != nil {
+		return err
+	}
+	mc.rootDir = workspaceDirFromFilePath(lockFile.Path)
+	w.moduleCache = &mc
+	w.logger.Printf("updated module cache - %d references parsed", len(mc.Modules))
+	return nil
+}
+
+func (w *workspace) Parser() lang.Parser {
+	w.pluginMu.RLock()
+	defer w.pluginMu.RUnlock()
+
+	return w.parser
+}
+
+func (w *workspace) ReferencesModulePath(path string) bool {
+	if w.moduleCache == nil {
+		return false
+	}
+
+	for _, m := range w.moduleCache.Modules {
+		if m.Key == "" {
+			// skip root module, as that's tracked separately
+			continue
+		}
+		if !strings.HasPrefix(m.Source, "."+string(os.PathSeparator)) {
```

Comment:
Not sure if this is accurate as I'm not sure what the non-relative paths look like in all cases, but this may be a better check:

```suggestion
		if filepath.IsAbs(m.Source) {
```

Reply:
Yeah - that's a good question - I need to try out a few scenarios to see what paths to expect in the cache.

Reply:
I added more explicit `IsExternal()` to the module record instead as that is cleaner and the logic there is also slightly different as this one was not as reliable - `Source` can also have `../` for modules which are local to some other modules.